Use unified Heroku buildpack output style

Malax opened the pull request.

schneems commented:
Not a blocker, but would help for QA-ing the output as a whole to print the getting started guide in CI.  https://github.com/heroku/buildpacks-ruby/blob/187770e700a433cf9e6af678cb3eb5735b89eec5/.github/workflows/ci.yml#L62-L97

schneems commented on shared/src/output.rs:
```diff
@@ -0,0 +1,365 @@
+
+pub fn print_section(text: impl Into<BuildpackOutputText>) {
+    let text = text.into().to_ansi_string();
+    println!("{ANSI_RESET_CODE}- {text}");
```
All code clearing needs to happen before the newline of the last print rather than at the beginning of the next line. Otherwise we end up with red/yellow/blue `remote:`-s.

schneems replied:
It won't hurt anything, I was just a little thrown off. 

I think what I really want is the output to look at. I said so earlier as a non-blocker, but I'll upgrade that to a strong request. I think that's more valuable than me making random comments about your comments i.e. I care more about the output than the implementation and need a way to view/iterate on it.

schneems commented on shared/src/output.rs:
```diff
@@ -0,0 +1,365 @@
+        let mut result = String::new();
+
+        // Every line must start with a style reset, the default ANSI code and the line prefix if
+        // it exists.
```
Needs to be the other way around. Each line must end with a reset code, but resume the prior code on the next line.

schneems replied:
I think this is for handling colorizing one value while inside of another? If so, maybe we update the comment to say why that's the case.

schneems commented on buildpacks/maven/src/main.rs:
```diff
@@ -201,30 +209,51 @@ impl Buildpack for MavenBuildpack {
+                command,
+                false,
+                MavenBuildpackError::MavenBuildIoError,
+                |output| MavenBuildpackError::MavenBuildUnexpectedExitCode(output.status),
```
I have some guidelines for running commands here https://github.com/heroku/buildpacks-ruby/pull/139. It seems that the output from the command being run won't be visible if this fails:

- [x] A representation of the command being run should appear in logs, or the error, or both
- [ ] Stdout and Stderr from the command should appear in logs, or the error, but NOT both (if it runs).
- [x] Make non-zero status an error
- [x] Allow for debugging a command that was run after execution (i.e., get command program name, arguments, env, cwd and pass them to something like WhichProblem)
- [x] Ideally only have to format Command once

Please update to include the output when this fails. I know some commands of yours produce absurd amount of data. If that's the case then maybe something like the last N lines of output? Basically what I'm after is: Why did this fail and what can I do about it? Having the status code is good, but I'm betting that there might be hints in the command output as well that will help.

Malax replied:
I'm not sure what the ask here is. Can you clarify? Currently, this outputs both `stdout` and `stderr` as it would on a local terminal. For Maven, the build error will be the last thing in the logs and Maven users are used to it. Are you suggesting adding the whole of `stderr` if the command fails, but filtering it out if it doesn't?

schneems replied:
> Currently, this outputs both stdout and stderr as it would on a local terminal. For Maven, the build error will be the last thing in the logs and Maven users are used to it. Are you suggesting adding the whole of stderr if the command fails, but filtering it out if it doesn't?

It's not outputting to stdout/stderr here. This code:

```rust
output::run_command(
    command,
    true,
    MavenBuildpackError::MavenBuildIoError,
    |output| MavenBuildpackError::MavenBuildUnexpectedExitCode(output.status),
)
```

Calls this branch where stdout and stderr are sent to `std::io::sink()` because quiet is true:

```
pub fn run_command<E, F: FnOnce(std::io::Error) -> E, F2: FnOnce(Output) -> E>(
    mut command: Command,
    quiet: bool,
    io_error_fn: F,
    exit_status_fn: F2,
) -> Result<Output, E> {
    let child = if quiet {
        command.output_and_write_streams(std::io::sink(), std::io::sink())
// ...
}
```

The docs for https://doc.rust-lang.org/beta/std/io/struct.Sink.html describe it as "A writer which will move data into the void." So it's never being seen by the user when quiet is set to true.

schneems replied:
I see the confusion now. I commented on the wrong line. The command invocation in question is L272, this one is clearly using the quiet = false argument. I had assumed the two maven commands had different error variants.

I'll close this comment and move it over to that line.

schneems commented on buildpacks/maven/src/main.rs:
```diff
@@ -238,10 +267,15 @@ impl Buildpack for MavenBuildpack {
+                MavenBuildpackError::MavenBuildIoError,
+                |output| MavenBuildpackError::MavenBuildUnexpectedExitCode(output.status),
+            )
+        })?;
```
Originally requested on the incorrect line https://github.com/heroku/buildpacks-jvm/pull/745#pullrequestreview-2476776597

Specific requests:

- Please either stream the output as it happens or add it to the error information (but not both). 
- Please add the command being run to the error as well (it doesn't look like it's emitted)
- If this command could take a variable amount of time (such as needing the network) please add background printing to let the user know we're still waiting for something to finish.

schneems replied:
I moved this feedback to a "lster" issue https://github.com/heroku/buildpacks-jvm/issues/762

schneems commented on buildpacks/maven/src/main.rs:
```diff
@@ -201,30 +209,51 @@ impl Buildpack for MavenBuildpack {
+                MavenBuildpackError::MavenBuildIoError,
+                |output| MavenBuildpackError::MavenBuildUnexpectedExitCode(output.status),
+            )
+        })?;
```
- There's a partial command emitted earlier but on failure we don't have the full command that was run, needed for debugging. Please add it or emit the entire command.

schneems replied:
I moved this feedback to a "lster" issue https://github.com/heroku/buildpacks-jvm/issues/762

schneems commented on buildpacks/maven/src/main.rs:
```diff
@@ -201,30 +209,51 @@ impl Buildpack for MavenBuildpack {
+                shell_words::join(&maven_options),
+                shell_words::join(&maven_goals)
+            )),
+        ]));
```
- Request: Log the command based off of what's going to be run. In this case the logging and the command can differ, and actually do here. You're running with `internal_maven_options` but are not printing them. Also you're using `maven_options.chain()` below but not here.

If that's intentional please comment it. Here's an example where I'm intentionally hiding command args from the user (though the full/real command will show on error/failure):

https://github.com/heroku/buildpacks-ruby/blob/187770e700a433cf9e6af678cb3eb5735b89eec5/buildpacks/ruby/src/layers/bundle_download_layer.rs#L99

schneems replied:
I moved this feedback to a "lster" issue https://github.com/heroku/buildpacks-jvm/issues/762

schneems commented on buildpacks/sbt/src/main.rs:
```diff
@@ -102,24 +103,30 @@ impl Buildpack for SbtBuildpack {
+            output::print_subsection(BuildpackOutputText::new(vec![
+                BuildpackOutputTextSection::regular("Running "),
+                BuildpackOutputTextSection::value(format!("sbt {}", shell_words::join(&tasks))),
+            ]));
```
You're printing exactly what you're going to run here (yay!). I still request to refactor based off of a single source of truth such that someone changing the command arguments won't accidentally diverge from what's being printed.

schneems replied:
I moved this feedback to a "lster" issue https://github.com/heroku/buildpacks-jvm/issues/762

runesoerensen commented on shared/src/output.rs:
```diff
@@ -0,0 +1,365 @@
+            match section {
+                BuildpackOutputTextSection::Regular(_) => {}
+                BuildpackOutputTextSection::Value(_) => {
+                    result.push_str(ANSI_VALUE_CODE);
+                    result.push(VALUE_DELIMITER_CHAR);
```
I think the spec calls for "value" types (e.g. `Value` and `Url`) to have color applied only to the value, and not the delimiter chars -- at least that's [how bullet_stream implements the spec](https://github.com/heroku-buildpacks/bullet_stream/blob/ec6a37f84c22d42f9a090bb8fbf45c3367373421/src/style.rs#L20-L21).

Seems like a sensible choice for CLI output?

Malax replied:
This has been fixed, thanks! :)

runesoerensen commented on shared/src/output.rs:
```diff
@@ -0,0 +1,365 @@
+                    result.push_str(ANSI_URL_CODE);
+                }
+                BuildpackOutputTextSection::Command(_) => {
+                    result.push_str(ANSI_COMMAND_CODE);
```
This should perhaps be wrapped in backticks (like `BuildpackOutputTextSection::Value`, but with this command color)?

schneems reviewed (changes requested):
Please merge https://github.com/heroku/buildpacks-jvm/pull/761, put this back in draft state and make changes until you're happy with the output (removal of newlines, using the correct style for a "value" etc.) Then we can iterate on the output style. Happy to iterate/pair too.

Malax commented on shared/src/output.rs:
```diff
@@ -0,0 +1,365 @@
+                }
+            }
+
+            if let BuildpackOutputTextSection::Value(_) = section {
```
Good catch, fixed in da00c86 and changed the code so that it will not compile if one section kind isn't handled.

Malax commented:
All (still applicable) comments have been addressed. :)

runesoerensen reviewed (approved):
Appreciate all your work on this! I've updated [my branch adopting this code](https://github.com/heroku/buildpacks-dotnet/pull/157) after your last tweaks, and it works great. I'm approving the PR, and just wanted to add some thoughts/comments (that are likely outside the scope of what needs to be addressed here):

* Would it make sense to always flush standard output/error at the end of the `output::print_*` functions? Or perhaps, more pragmatically, just calling `io::stdout().flush()` at the beginning of the two functions that write to `stderr` (i.e. `output::print_error` and `output::print_warning`)?

    There's a risk the output isn't displayed in the "expected" order when writing to both streams, e.g. this:

        output::print_section("Foo");
        output::print_error("Bar", "Baz");

    Should probably print output like this:

        - Foo
        ! ERROR: Bar
        ! 
        ! Baz

    But very frequently, the standard error output will appear before the standard output, e.g.:

        ! ERROR: Bar
        ! 
        ! Baz
        - Foo

    I'm curious if you have thoughts on this? It's of course fairly trivial to explicitly flush where needed, but should this be necessary when using the `output` module (as opposed to directly interacting with the standard error/output streams)?
* The output from the Gradle buildpack could perhaps be improved. It's a bit tricky dealing with a daemon running in the background, and that situation hasn't changed with this PR as far as I can tell, but the unindented output from the daemon sticks out a bit (now that the buildpack output is indented/using sections). Would it make sense to indent the output to produce something like this?
```
- Running Gradle build
  - Starting Gradle daemon
      Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF-8
      Downloading https://services.gradle.org/distributions/gradle-8.12-bin.zip
      .............10%.............20%.............30%.............40%.............   50%.............            60%.............70%.............80%.............90%.............100%
      Starting a Gradle Daemon (subsequent builds will be faster)
      > Task :heroku_buildpack_start_daemon UP-TO-DATE
      
      BUILD SUCCESSFUL in 30s
        - Done (31.2s)
        - Querying tasks
        - Done (1.4s)
        - Querying dependency report
        - Done (2.9s)
- Running Gradle build
  - Running `./gradlew build -x check`
      Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF-8
      > Task :compileJava
      > Task :processResources
      > Task :classes
      > Task :resolveMainClassName
      > Task :bootJar
      > Task :jar
      > Task :assemble
      > Task :build
```
* On a related note, it seems a bit off that there are now two sections with the same text, `Running Gradle build`. I noticed the old headers were changed in this PR, perhaps to accomodate the section/subsection style (or maybe I'm missing something) -- but would it make sense to name the sections something like "Gradle Daemon" and "Gradle Build", and printing the commands as subsections?
* It's great that `output::print_warning` and `output::print_error` include the `!` prefix on all lines in the body. However, empty lines have a trailing whitespace/space after the `!`.

    I can see some merit to using the same prefix for all lines in the error output, but guess I'd prefer it if empty error/warning body lines were written as `!` rather than `! `.
    
    This is definitely a nitpick and likely not worth the added complexity to implement. Just wanted to point out this difference compared to for instance the bullet_stream implementation ([here's the relevant diff](https://github.com/heroku/buildpacks-dotnet/pull/157/commits/6a6372c8af0de3c4c044bf92bb127eef554bf896#diff-7b4cacf7af98ed35441ecfa0fbef6121440a826ee5469a42054e555a668a68acR44-R53)).

schneems reviewed (changes requested):
## SBT buildpack output

```
# Heroku sbt Buildpack
```

H2 header for buildpack names (all buildpacks)

```
      [info] Non-compiled module 'compiler-bridge_2.13' for Scala 2.13.10. Compiling...
      [info]   Compilation completed in 9.87s.
      [info] done compiling
      [success] Total time: 15 s, completed Jan 17, 2025, 2:45:49 PM
      [info] Wrote /workspace/target/scala-2.13/scala-getting-started_2.13-1.0-SNAPSHOT.pom
      [success] Total time: 1 s, completed Jan 17, 2025, 2:45:49 PM

  - Done (48.6s)
```

Missing some kind of a global "done" to separate the output of this buildpack from future buildpacks. (all buildpacks)


## Gradle buildpack

Indentation on this streaming command:

```
- Running Gradle build
  - Starting Gradle daemon
Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF-8
Downloading https://services.gradle.org/distributions/gradle-8.12-bin.zip
.............10%.............20%.............30%.............40%.............50%.............60%.............70%.............80%.............90%.............100%
Starting a Gradle Daemon (subsequent builds will be faster)
> Task :heroku_buildpack_start_daemon UP-TO-DATE
```

schneems reviewed (changes requested):
~~We are getting double warning output when no version is specified, one in the old format and one in the new:~~

Correction it's a different warning, but we're still mixing and matching log styles:

```
$ cargo test --locked -- --ignored --test-threads 16 openjdk_default
! WARNING: No OpenJDK version specified
!
! Your application does not explicitly specify an OpenJDK version. The latest
! long-term support (LTS) version will be installed. This currently is OpenJDK 21.
!
! This default version will change when a new LTS version is released. Your
! application might fail to build with the new version. We recommend explicitly
! setting the required OpenJDK version for your application.
!
! To set the OpenJDK version, add or edit the system.properties file in the root
! directory of your application to contain:
!
! java.runtime.version = 21

[Warning: Using default version]
Your application does not explicitly specify which Maven version should be used to build your application.
The current default version 3.9.4 will be used. This default version will change from time to time.
Depending on your build configuration, a different Maven version might cause your build to fail.

We recommend that you use Maven wrapper to ensure that the same Maven version is always used to build your
application. Alternatively, you can set 'maven.version' in 'system.properties' to a supported Maven version.

thread 'main' panicked at buildpacks/maven/src/warnings.rs:28:5:
Nope
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
ERROR: failed to build: executing lifecycle: failed with status code: 51
```

From here https://github.com/heroku/buildpacks-jvm/blob/3b140fce22ba8f186f393676fb3f02acfee03b68/buildpacks/maven/src/warnings.rs

runesoerensen commented:
> H2 header for buildpack names (all buildpacks)

Definitely agree we should use the same buildpack name headers across all buildpacks, but the spec actually calls for h1 headers. Maybe that needs to be updated to reflect how it's implemented in `bullet_stream`/should be implemented here?

schneems commented:
@runesoerensen I told him to do it. I opened https://github.com/heroku-buildpacks/bullet_stream/issues/23 to bring that to the larger group.

schneems commented:
Here's my take on this PR. It's implemented on top of this branch. https://github.com/heroku/buildpacks-jvm/pull/767 

<img width="152" alt="image" src="https://github.com/user-attachments/assets/19aab5e5-a7ec-483d-a53d-575e0c3c0457" />

Take anything that's useful...drop anything that's not. Doing that helped me find some missing changes such as left-over `libherokubuildpack::log` calls. It prints more command info to the logs, but IMHO that's fine unless there's strong reasons to hide one. I also refactored the command error handling which would close out #762.

schneems reviewed (changes requested):
## Known issues

These are things need to be done to migrate onto the new output format, there might be more, but all of these recommendations come from having implemented the migration twice already. These almost exclusively come from specific commits on this ready-to-merge PR https://github.com/heroku/buildpacks-jvm/pull/769:

- [x] Remove `libherokubuildpack::log` references and the feature from:
  - [x] Gradle (https://github.com/heroku/buildpacks-jvm/pull/769/files#diff-07f05f536de71a214fed08fa6be322b2bc41fde291d9be0cf668c2e3440638ddR13)
  - [x] JVM (https://github.com/heroku/buildpacks-jvm/pull/769/files#diff-e3cbda9372a9cc72f97e8c9e2513c7836a912b2e722864536a8e1997a18ee46aR14)
  - [x] Shared (https://github.com/heroku/buildpacks-jvm/pull/769/files#diff-401919e5af595c40b1359c42a25b4c698681b191f33b5b0ceba9ab87d40ce022R12)
  - [ ] Maybe JVM function invoker (Your call)
- [x] Fix tests after removing references to `libherokubuildpack::log`
- [x] Make Buildpack name an H2 (https://github.com/heroku/buildpacks-jvm/pull/769/files#diff-899c56e56432daa5771452ed91965be83b0ee62a30145bed0702b4c77a761a59R8)
- [x] Add a "this buildpack finished" output for all buildpacks, examples:
  - https://github.com/heroku/buildpacks-jvm/pull/769/commits/91becf025b272b3aa4bed74b337e892747e8deaa
  - https://github.com/heroku/buildpacks-jvm/pull/769/commits/62900adb6eb899aec8616f107adc0e3f340ea69c
- [x] Quiet/remove any accidental printing to stdout/stderr https://github.com/heroku/buildpacks-jvm/pull/769/commits/cdb3e5116890985bee02bd22c5c9a86868dd01f0
- [x] Fix gradle "-x check" command execution format https://github.com/heroku/buildpacks-jvm/pull/769/commits/dfd5504bf39ef6ce58d918f14cdf891569adba7d

## Double check

While iterating, I originally made changes and looked at the print output and kept going. The above :up: list is based on the second time I implemented the build output for buildpacks-jvm and is more focused, however, it's important that the output be used while iterating or you might introduce different problems. Basically "it's not done until it looks okay."

- [x] Timing information isn't being emitted at the same time as commands are streaming
- [ ] Command execution indentation and formatting is correct
- [ ] Commands "Running: `<cmd>`" should be indented under an appropriate header
- [ ] The "print" CI task's output looks okay in all three guides

## Known Missing

We can move forwards without these things, but they're known "need to implement it" eventually. If we don't do it now, we can move to an issue (or some are already issues).

- [ ] Emit a background timer for commands (and tasks). Examples:
  - https://github.com/heroku-buildpacks/bullet_stream/blob/7a057577b560fa5ed0e346943bbc4dcd004773fe/src/global.rs#L228
  - https://github.com/heroku-buildpacks/bullet_stream/blob/7a057577b560fa5ed0e346943bbc4dcd004773fe/src/global.rs#L249
  - https://github.com/heroku-buildpacks/bullet_stream/blob/7a057577b560fa5ed0e346943bbc4dcd004773fe/src/lib.rs#L602C1
  - https://github.com/heroku-buildpacks/bullet_stream/blob/7a057577b560fa5ed0e346943bbc4dcd004773fe/src/lib.rs#L630
- [ ] Emit the exact command being run. On error always make sure output is available for debugging: https://github.com/heroku/buildpacks-jvm/issues/762
- [ ] All cache invalidation is visible to the end user along with an explanation for why it was invalidated

## Bonus

Not required but makes life nicer:

- [ ] (optional) Let me see which CI tasks are associated with which guides without needing to click https://github.com/heroku/buildpacks-jvm/pull/769/commits/02c06f3e6bfeee948714385ed19086bf7f75cf51

schneems commented:
To round back on the h2 issue, I went to send a PR to the dotnet buildpack and it looks like it is already using them https://github.com/heroku/buildpacks-dotnet/blame/f11d9b15c3ae79a2d364a68b42f4973b05ee90e8/buildpacks/dotnet/src/main.rs#L68.

schneems reviewed (changes requested):
I can confirm that the specific fixes I requested have been implemented. After reviewing the printed buildpack output, I found some issues, which are detailed below.

## Required fixes

- [x] Maven dependency list is not indented, 

```
- Running `./mvnw dependency:list` quietly
  - Done (4.0s)
- Done (finished in 10.7s)
```

should be

```
  - Running `./mvnw dependency:list` quietly
  - Done (4.0s)
- Done (finished in 10.7s)
```

From the current context, it's unclear why it's being run. It's under the `- Running Maven build` bullet point. It could be under another bullet point, but I'm not sure what. What's the higher order reason you're running that command? Looking at your code, it seems the `detect_framework` function consumes it. So I would suggest:

```
- Java Framework
  - Running `./mvnw dependency:list` quietly
  - Done (4.0s)
  - Detected Spring Boot (`org.springframework.boot`)
```

```
- Java Framework
  - Running `./mvnw dependency:list` quietly
  - Done (4.0s)
  - Detected Wildfly Swarm (`org.wildfly.swarm`)
```

```
- Java Framework
  - Running `./mvnw dependency:list` quietly
  - Done (4.0s)
  - None detected (did not find `org.springframework.boot` or `org.wildfly.swarm`)
```

- [x] Gradle daemon command output needs to be formatted or hidden

```
## Heroku Gradle Buildpack

- Running Gradle build
  - Starting Gradle daemon
Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF-8
Downloading https://services.gradle.org/distributions/gradle-8.12-bin.zip
.............10%.............20%.............30%.............40%.............50%.............60%.............70%.............80%.............90%.............100%
Starting a Gradle Daemon (subsequent builds will be faster)
> Task :heroku_buildpack_start_daemon UP-TO-DATE

BUILD SUCCESSFUL in 31s
  - Done (31.6s)
  - Querying tasks
  - Done (1.4s)
  - Querying dependency report
  - Done (4.0s)
```

Given that it takes so long (31s), I recommend streaming it as a normal `Running <cmd>` output. That's different from what you're currently doing but less confusing. If you were using `bullet_stream,` I would suggest background dots here as an alternative, i.e., "Starting Gradle daemon ... "


- [x] Missing "done" timing on the command execution:

```
- Running Gradle build
  - Running `./gradlew build -x check`

      Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF-8
      > Task :compileJava
      > Task :processResources
      > Task :classes
      > Task :resolveMainClassName
      > Task :bootJar
      > Task :jar
      > Task :assemble
      > Task :build
      
      BUILD SUCCESSFUL in 6s
      5 actionable tasks: 5 executed

- Done (finished in 44.5s)
```

Should be:

```
      BUILD SUCCESSFUL in 6s
      5 actionable tasks: 5 executed

  - Done (2s)
- Done (finished in 44.5s)
```

From an interface perspective, every time you need to run a streaming command, you also need to remember to emit "Done" with timing. My suggestion is to couple that behavior with your implementation. i.e. have `run_command` emit it by default so you don't forget it.

## Optional/Internal suggestions

These aren't strictly required. My suggestions relate to your public code interfaces. You can choose to ignore them (hence the optional tags), but these also relate to the design space I've been playing in. These all relate to running and displaying commands:

Short: I recommend you couple printing the name of the command, streaming the command, and printing is done into one interface. I suggest this interface only takes a single command input, and you don't try to hide the `-B` of maven commands. Return a custom error that can be `map_err`-d instead of requiring two functions as inputs. For running commands quietly, I suggest either introducing a different interface (that also couples advertisement) or providing no interface and requiring someone to manually call `command.output()` while annotating using the existing print and timing functionality.

Long (individual changes and why I'm suggesting them):

- [ ] (Optional) Change the boolean argument of `run_command` to something else or remove that functionality from that interface. 

When I was implementing my PRs, the boolean flag for `run_command` was a little confusing. I kept having to look at the argument annotation from my editor and say out loud in my head "quiet false means show to stream the output". Taking an enum could be more clear, for example:

```
run_command(command, output::RunCmd::Quiet, ...)
run_command(command, output::RunCmd::Stream, ...)
```

Alternatively, The "quiet" branch doesn't do anything (other than enforce that the exit status has been checked). This function is in the `output` module, but calling it doesn't output anything. That option could be removed. 

You could also simplify the interface by returning an enum like `CommandError` with variants `NonZeroExit` and `SystemError` [like this](https://github.com/schneems/fun_run/blob/496a6b8b2d17e396b729e512836f4eb6f5a4c728/src/lib.rs#L419), then you let your caller `map_err` and decide what to do with it (or treat both as a failed command depending on your use case). This approach is more ergonomic than the accepting callback functions in your interface.

- [ ] (Optional) Couple printing "Running <cmd>" with streaming command execution (in addition to auto adding the `- Done` timing I mentioned above). 

Streaming a command requires three things: Advertising the command being run, streaming the output, and advertising that the command is done. These are currenly exposed in your library as three separate features you must remember to add in order to get the right output. They are tightly coupled...so couple them in the code. Alternatively, if you don't want to do that, you could make it clearer that `run_command` is more about formatting the stream with renaming.

One guess related to why you're resisting this coupling is related to an observation I made when working on my PRs: you seem to want to hide the `-B` when executing the commands. Even if it's a little odd to users (guessing that's why you did it) it's the actual command we're running. Printing the exact command we're running is a good default, and modifying a command's representation should generally be used to add information rather than obfuscate what we're actually doing. It's fine. I also see `assert_not_contains` against `-DoutputFile=target/mvn-dependency-list.log`, which also seems okay.

If you wanted to preserve the ability to run with renaming, you could make the function signature `announce_stream_command(command_name: String, command: Command) -> CommandError` (for example). I suggest YAGNI most of the time. If you're willing to rely on a 3rd party crate interface you could take the trait `announce_stream_command(command: impl fun_run::CommandWithName)`. An alternative would be for you to introduce your own trait. The downside is it requires you to import it everywhere you want to use it, but at least that's compile time checked.

Malax commented:
The Maven output seems fine to me in the context of this PR. Let's stick to moving the output style to the new one before making fixes to the output/UX in general. Indentation also seems correct to me on CI. I left this as-is.

The other issues I have fixed. I found another one with duplicated Gradle headings that I've fixed too.

Everything optional I left out to be either broken out or discussed separately.

schneems reviewed (changes requested):
Re-requesting indenting this line:

```
- Running `./mvnw dependency:list` quietly
```

It's fixed via this stacked PR (so you can see the CI print output) if you just wanna hit the big merge button. Other than that looks good to go. 

Full:

----

>  Missing "done" timing on the command execution:

Confirmed done ✅ 

>  Gradle daemon command output needs to be formatted or hidden

Confirmed done ✅ 

>  Maven dependency list is not indented,

Not done, following up on comment:

> The Maven output seems fine to me in the context of this PR. Let's stick to moving the output style to the new one before making fixes to the output/UX in general. Indentation also seems correct to me on CI. I left this as-is.

This is not currently up to spec/standard. This output it's emitting now:

```
  - Done (3.3s)
- Running `./mvnw dependency:list` quietly
  - Done (2.0s)
- Done (finished in 5.3s)
```

Which isn't possible with the reference implementation in strict mode (unless you wanted to hand-roll some stuff and try really hard). 

This is what I asked for as a minimum (indenting the command):

```
  - Running `./mvnw dependency:list` quietly
  - Done (4.0s)
- Done (finished in 10.7s)
```

I'm trying to be pragmatic by spelling out what I **actually** want as well as the minimum i'm willing to accept. This current output doesn't meet my minimum. Can you clarify your pushback on why you didn't want to indent it, or were you thinking I was saying my minimum was to add a new header?

Technically even that's not up to standard as your implemention doesn't emit dots. It's not possible to emit a timed Done like you're doing here. But that's nitpick I don't feel like pushing right now.

## Optional

- [ ] (Optional) Buildpack timing output as a struct instead of a function: Another implementation note. I feel the buildpack timing function would be better not taking a funciton as an argument, but rather that it return a struct that must be used. The current implementation forces an indentation which makes diffs harder to compare.

schneems commented:
I'm leaving for vacation. I want to hand this off to another reviewer so I'm trying to be explicit with my merge guidelines:

I see no reason (and have gotten no replies) for why #776 is a problem. If that goes in (or something similar) then I'm ready to merge, like a week ago. If there's a reasonable hesitation or debate about that line and Manuel wants to document that and move it into an issue for discussion and merge this now: I'm okay with that provided there's actual movement and an intent to come to amutual understanding and move forward productively.

If a new reviewer wants to add anything that's fine, but it's not great to have long lived branches and this has been open since November. I see no reason why this can't be merged and deployed.

schneems reviewed (approved):
Approved after bullet -> sub-bullet suggestion is accepted.

schneems commented:
Can you rebase this please?
